### ISSUE TITLE
ALTO output from Abbyy11r8 contains <SP> between <String>

### ISSUE THREAD

**Issue by jbarth-ubhd** [UB-Mannheim/ocr-fileformat#78]
Perhaps this is not an error.
Kind regards,
J. Barth

**Comment by kba** [comment]
Can you provide sample data and how you ran the tool?

**Comment by zuphilip** [comment]
I guess you output the ALTO files directly from ABBYY, because we don't yet provide a transormation from ABBYY to ALTO. Then this should be an example: https://digi.bib.uni-mannheim.de/~stweil/ocr-praxis/Testseiten/alto/417576986_0031.xml

The `<SP>` stands AFAIK for space and it should validate in this form.

**Comment by jbarth-ubhd** [comment]
Yes, I'll try to find out if <SP> (=space) is really necessary between <String>s in ALTO.

**Comment by zuphilip** [comment]
I guess that it still validates without the SP tags. Moreover, most of the information (HPOS, WIDTH) can be calculated from the line above and below, but if the width of a space is important for some application, then it might be easier to have this data directly. I don't know what the VPOS information for a space says or whether it is also determined by some other values.

**Comment by jbarth-ubhd** [comment]
On ALTO 2.1 .xsd it looks like this:

```
  \<xsd:sequence maxOccurs="unbounded"\>
    \<xsd:element name="String" type="StringType"/\>
    \<xsd:element name="SP" minOccurs="0"\> ...
    \</xsd:element\>
  \</xsd:sequence\>

```

So strictly speaking it seems that \<SP\> is not necessary, but the \<sequence\> seems to imply it.

**Comment by zuphilip** [comment]
> but the <sequence> seems to imply it.

Not sure. I only see here, that, if `<SP>` occurs, then it has to occur after a `<String>`.

**Comment by stweil** [comment]
The [ALTO documentation](https://www.loc.gov/standards/alto/techcenter/layout.html) says "A TextBlock is divided into lines and those are divided into strings, spaces and hyphens". I don't interpret that as a strict requirement that spaces are required, and nor does the .xsd. It's clear that spaces are required if the strings are given without `HPOS` and `WIDTH` attributes, but I think it is redundant if those attributes are available.

**Comment by amitdo** [comment]
The ALTO spec itself needs to clarify this issue.

**Comment by stweil** [comment]
Clemens has created an issue for that: https://github.com/altoxml/schema/issues/54 (thank you).

**Comment by cneud** [comment]
Thanks for flagging this, I will put it on the agenda for our next ALTO board call which will be held November 29th.

**Comment by mittagessen** [comment]
To chip in, I've interpreted the standard that the `<SP><String>` alternation is mandatory and that whitespace should never occur inside a `<String>` and this is how I implemented it.

**Comment by stweil** [comment]
If a `<String>` never contains whitespace, then `<SP>` is completely redundant. Does ALTO allow overlapping words in a row? If yes, does that require a separating space with negative width? :-)

**Comment by mittagessen** [comment]
> If a <String> never contains whitespace, then <SP> is completely redundant.

Why? Whitespace is a character like any other and personally I would've taken the decision to encode it explicitly using `<String>` if the standard wouldn't heavily imply that you shouldn't do that. Of course, you can throw away the data and let people compute inter-word spacing implicitly provided through word bounding boxes but it isn't like tesseract, kraken or any other sequence classification based OCR engine doesn't output a label for whitespace (and the boundaries of that activation can almost certainly differ from the boundary of the activations of the adjacent letters). I'd rather not throw away metadata that some weird subdiscipline in the humanities that only the 8 people participating in it have ever heard about might need.

> Does ALTO allow overlapping words in a row?

ALTO luckily allows overlapping elements in constrast to PageXML.

**Comment by stweil** [comment]
Then how would you encode two overlapping words if you are forced to put a `<SP>` between them?

**Comment by mittagessen** [comment]
Just have overlapping bounding boxes? Presumably there is still a reading order that determines the ordering of the `<String>` tags. But yeah it helps that I decided a long time ago that words are a waaay to squishy concept and arbitrarily defined anything bounded by whitespace is a separate word/segment for serialization purposes (not only for ALTO). Of course, I you want to encode a proper tokenization, this data model shouldn't be used. On the other hand, I'm of firm conviction starting to do that in a raw OCR serialization format is only going to lead to madness.

**Comment by cneud** [comment]
Just to follow up - I'm afraid a quick resolve is not really around the corner...the issue was discussed in the last ALTO board call, with the core elements of the discussion summarized [here](https://github.com/altoxml/schema/issues/54).

While the general feeling was that the use of `<SP>` is not mandatory, some more research into ALTO's history is required to determine the original authors exact intentions.

An expansion of the `<SP>` tag with a `width` attribute has been identified among board members as a possibility to create more useful future applications for the `<SP>` tag.

If one really wants to be on the safe side, the quick solution right now would be to indeed include `<SP>` in the output of any ALTO export implementation as it is also straightforward to remove in post-processing.

**Comment by mittagessen** [comment]
> ALTO's history is required to determine the original authors exact intentions.

As a note, most of the character-based classification systems common at the time ALTO was originally specified didn't treat whitespace as a proper glyph, i.e. whitespace is just something bordered by other glyphs and is never seen by the classifier as such. This at least explains the existence of a separate `<SP>` tag.

**Comment by stweil** [comment]
Thank you, @cneud, @mittagessen and the ALTO board.

As the current DFG viewer expects the `<SP>` tags, I think that programs like `ocr-transform` should produce them, too. Pull request https://github.com/tesseract-ocr/tesseract/pull/2117 adds the tags to Tesseract's new ALTO output, so that output is now compatible with the DFG viewer.

**Comment by zuphilip** [comment]
The addition of the `<SP>` should be handled uspream in the corresponding transformation. Currently, we use hocr2alto and page2alto. We can keep this issue here open as a reminder.

**Comment by filak** [comment]
According to the ALTO XSD the SP tag is optional - minOccurs="0"

And I do not see a way how to reliably calculate HEIGHT/WIDTH/VPOS/HPOS attributes from the hOCR data for the SP tag. 

IMHO - proper handling of optional SP tag should be fixed by DFG viewer.

**Comment by albig** [comment]
If the `<SP>` is not mandatory, we have to "ignore" it in the styles of the fulltext view and always make a space after a `<STRING>`. 

This is what I've done in the DFG-Viewer styles now. Please have a look at the current master of the DFG-Viewer at [test.dfg-viewer.de](https://test.dfg-viewer.de/). 

Please compare the example from above in [current master](https://test.dfg-viewer.de/show/?id=8071&tx_dlf%5Bid%5D=https%3A%2F%2Fdigi.bib.uni-mannheim.de%2Foai%2F%3Fverb%3DGetRecord%26metadataPrefix%3Dmets%26identifier%3D511144768&tx_dlf%5Bpage%5D=8) and in [version 5.0](https://dfg-viewer.de/show/?id=8071&tx_dlf%5Bid%5D=https%3A%2F%2Fdigi.bib.uni-mannheim.de%2Foai%2F%3Fverb%3DGetRecord%26metadataPrefix%3Dmets%26identifier%3D511144768&tx_dlf%5Bpage%5D=8) of DFG-Viewer and report change requests.

**Comment by filak** [comment]
@albig IMHO the second one seems better from user perspective - it is more readable/compact and if you need to copy it and ie. quote the text in your paper/text you will need to remove the blanks manually.

**Comment by sebastian-meyer** [comment]
@albig IMHO the spacing looks better now (in master), but the linebreaks seem a bit random...
